lib/mobility/backends/active_record/table.rb

```diff
@@ -137,6 +137,14 @@ def self.configure(options)
           class_name:  name,
           foreign_key: options[:foreign_key],
           inverse_of:  association_name
+
+          callback_methods = Module.new do
+            define_method :initialize_dup do |source|
+              super(source)
+              self.send("#{association_name}=", source.send(association_name).map(&:dup))
+            end
+          end
+          include callback_methods
```

Comment:
This is in general the right way to do this, the only thing I'd point out is that if you define your attributes separately, e.g.:

```ruby
class Post < ApplicationRecord
  extend Mobility
  translates :title
  translates :content
  translates :author
end
```

Then this method will actually `super` through the same code three times. This might seem a bit pedantic, but in Mobility I try to ensure that defining attributes separately or together essentially does the same thing, so try to avoid this type of thing (even where the result, like here, is the same, since assigning the same thing any number times does not change anything).

This code path can be avoided by defining a constant for the module which includes the (camel-cased) name of the association as well as the backend, and then including that:

```ruby
module_name = "MobilityTable#{association_name.camelcase}"
unless const_defined?(module_name) do
  callback_methods = Module.new do
    # ...
  end
  include const_set(module_name, callback_methods)
end
```

This way, even if you call the code multiple times, you will only include the module once (unless the association name is different, in which case you should include it separately anyway.

There may be a simpler/better way to do this...

Reply:
Yeah, your approach is a good way of doing it. Strictly speaking, I don't think you need the `const_defined?` check, as including the same module multiple times won't do anything, but I guess it is safer to keep the check.

The other approach would be to have setup include a single module, that generates an initialize_dup method that handles associations. Something like

``` ruby
setup do |_attributes, options|
  self.mobility_association_names << association_name
  include InitializeDup
end

module InitializeDup
  def initialize_dup(source)
    super
    self.class.mobility_association_names.each do |association_name|
      self.send("#{association_name}=", source.send(association_name).map(&:dup))
    end
  end
end
```

I gather though you've designed mobility with avoiding adding such state to the classes it uses it.

Reply:
> Strictly speaking, I don't think you need the const_defined? check, as including the same module multiple times won't do anything, but I guess it is safer to keep the check.

Yes that's true, I just thought if it's already defined there's no point re-defining it - but agree better without since it saves a couple lines which only apply to an edge case anyway.

> I gather though you've designed mobility with avoiding adding such state to the classes it uses it.

Yes exactly, the goal is to add as little to the model class as possible, and a name like `mobility_association_names` would potentially conflict with other backends (which could be use din the same model).

It's overkill for 99% of *current* use-cases, but I want to leave the door open to use any combination of backends, applied any number of times, in the same model without fear that they would conflict with either each other in any way.

The tradeoff is that you have to define a constant on the model. I'm happy with any constant name as long as it includes the backend name and the associaiton name.